File: java/leetcode/src/main/java/com/leetcode/BST/Find_Mode_in_Binary_Search_Tree_501.java
package com.leetcode.BST;

import java.util.ArrayList;
import java.util.List;

public class Find_Mode_in_Binary_Search_Tree_501 {
    int maxCnt = 0;
    int cnt = 0;
    TreeNode pre = null;
    List<Integer> modes = new ArrayList<>();

    public int[] findMode(TreeNode root) {
        dfs(root);
        update();
        return getRes();
    }

    int[] getRes() {
        int[] res = new int[modes.size()];
        int i = 0;
        for (int m : modes) {
            res[i++] = m;
        }
        return res;
    }

    void dfs(TreeNode node) {
        if (node == null) return;
        dfs(node.left);
        if (pre == null) {
            cnt = 1;
        } else {
            if (pre.val == node.val) {
                cnt++;
            } else {
                update();
            }
        }
        pre = node;
        dfs(node.right);

    }

    void update() {
        if (cnt == maxCnt) {
            modes.add(pre.val);
            cnt = 1;
        } else if (cnt > maxCnt) {
            modes.clear();
            modes.add(pre.val);
            maxCnt = cnt;
            cnt = 1;
        } else {
            cnt = 1;
        }
    }
}

File: js/algrithm/huahua130/BST/Find_Mode_in_Binary_Search_Tree_501.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root) {
    let cnt = 0
    let max = 0
    let pre = null
    let modes = []

    const dfs = function(node) {
        if (node === null) return
        dfs(node.left)
        cnt = node.val === pre?.val ? cnt + 1: 1
        if (cnt === max) {
            modes.push(node.val)
        } else if (cnt > max) {
            max = cnt
            modes = [node.val]
        }
        pre = node
        dfs(node.right)
    }

    dfs(root)

    return modes
};